odoo/tools/pdf.py

```diff
@@ -19,3 +49,302 @@ def merge_pdf(pdf_data):
+        """
+        super().__init__()
+        self.pdf_content = b""
+        self.stack = []
+        if pdf_content is not None:
+            self.from_binary(pdf_content)
+
+    def from_binary(self, pdf_content):
+        """
+        Initialise the writer with the content from the provided bytes
+        :param pdf_content: the pdf content in bytes
+        """
+        self.pdf_content = pdf_content
+        reader_buffer = io.BytesIO(pdf_content)
+        reader = PdfFileReader(reader_buffer)
+        self.cloneDocumentFromReader(reader)
+        reader_buffer.close()
+
+    def make_pdfa(self, file_title="PDF"):
+        """
+        Take the content of a pdf generated by wkhtmltopdf and use ghostscript to make PDF/A-3B compliant.
+        :param file_title: the title of the file, used to set optional docinfo in the pdf
+        :return: a PDF/A-3B compliant file
+        """
+        # Replace the DOCINFO as wkhtmltopdf export one with utf-16be characters and ghostscript doesn't like it.
+        info = self.getObject(self._info)
+        info.update({
+            NameObject("/Creator"): createStringObject("Odoo"),
+            NameObject("/Title"): createStringObject("Invoice")
+        })
+
+        gs_bin = _get_ghostscript_bin()
+        pdfa_content = self.save_content()
+        if gs_bin:
+            try:
+                cmd = [
+                    gs_bin,
+                    '-dQUIET',  # Remove all routine information from the output
+                    '-dPDFA=3',  # Generate a PDF/A-3 file
+                    '-dBATCH',
+                    '-dNOPAUSE',
+                    '-dSAFER',  # Using DSAFER disallow access to the files on the system.
+                    '-sColorConversionStrategy=UseDeviceIndependentColor',
+                    '-sProcessColorModel=DeviceRGB',
+                    '-sDEVICE=pdfwrite',
+                    '-dPDFACompatibilityPolicy=1',  # Discard features that would result in a not PDF/A compliant file
+                    '-sOutputFile=-',  # "-" as output file makes it write to stdout instead of a file
+                    '-',  # "-" as input file is required since we provide the data from stdin
+                ]
+                # Run the command in a new subprocess. Use communicate to send the file input, and get the output
+                process = Popen(cmd, stdin=PIPE, stderr=PIPE, stdout=PIPE)
+                pdfa_content, error = process.communicate(pdfa_content)
+
+                if error:
+                    _logger.error("Error while converting PDF to PDF/A: %s", error)
+                    pdfa_content = self.pdf_content
+            except FileNotFoundError as error:
+                _logger.error("Error while calling Ghostscript: %s", error)
+                pdfa_content = self.pdf_content
+
+        # "Reset" the pdf writer with the output of ghostscript
+        self.__init__(pdfa_content)
```

Comment:
This `__init__` is dangerous and makes me think it might not be a good idea to handle the PDF-A inside the OdooPdfFileWriter: TO BE DISCUSSED